### ISSUE TITLE
demo feature for Karaf

### ISSUE THREAD

**Issue by maggu2810** [openhab/openhab2#375]
The demo configuration and bindings should be added to a new Karaf feature.

This is related to the broken openhab2-demo distribution in the __Karaf branch__ after merging maggu2810#1

**Comment by maggu2810** [comment]
@kaikreuzer Can you add the label "karaf"?

**Comment by kaikreuzer** [comment]
I will have a try to define such a feature.

**Comment by kaikreuzer** [comment]
Implemented with https://github.com/openhab/openhab2/pull/425

**Comment by maggu2810** [comment]
Short question before having a look at your PR:
Have you already considered to use a custom distribution for the demo?
I do not know if the demo is a feature someone want to (un)install in a runtime.
ATM there is a separate "application" - the openhab2 demo - that could be used to inspect openhab und the runtime + addons.
The addons have been implemented using features.
Do you think for the demo is a feature the better choice (instead of a custom distribution)?

**Comment by kaikreuzer** [comment]
I actually like the idea of just having a single downloadable. Actually, in openHAB 1, the demo.zip only contains the demo files and you extract them over the runtime.zip - so pretty similar as installing a feature. Many people also take the demo as a starting point and evolve it to their production system, so it is usually not a "play with demo" and then "start from scratch with the runtime".

I am not yet decided on how to best offer the installation of the demo. Asking people to do it manually on the shell is not my plan. Rather have a welcome screen on the dashboard offering to install the demo or something like this.

What do you think?

**Comment by kaikreuzer** [comment]
Do you have any idea why Karaf restarts many bundles upon feature installation? The demo feature install takes quite long and even single bindings seem to take a while due to a lot of re-resolution/restarting...

**Comment by maggu2810** [comment]
Will try to find some time to test it.
The PR does not build. Will try to cherry pick 35dd22f005fb371137ded49ad5e4d6d922a69e80

**Comment by maggu2810** [comment]
On my laptop the feature installation is fast (but I think it has enough power at all)...
But I do not see any restarts in the logs. Perhaps I have to increase logging.

I see an exception thrown by the yahoo weather binding
<pre>
2015-10-15 23:09:26,658 | INFO  | pool-28-thread-1 | FeaturesServiceImpl              | 9 - org.apache.karaf.features.core - 4.0.2 |   org.eclipse.smarthome.binding.yahooweather/0.8.0.201510151608
2015-10-15 23:09:26,723 | INFO  | ESH-safeCall-2   | ThingAddedEvent                  | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | Thing 'yahooweather:weather:berlin' has been added.
2015-10-15 23:09:26,724 | INFO  | ESH-safeCall-2   | ThingStatusInfoEvent             | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' updated: INITIALIZING
2015-10-15 23:09:26,725 | INFO  | ESH-safeCall-2   | ThingStatusInfoChangedEvent      | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' changed from UNINITIALIZED to I
NITIALIZING
2015-10-15 23:09:26,762 | INFO  | ESH-safeCall-2   | ThingStatusInfoEvent             | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' updated: ONLINE
2015-10-15 23:09:26,763 | INFO  | ESH-safeCall-2   | ThingStatusInfoChangedEvent      | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' changed from INITIALIZING to ON
LINE
2015-10-15 23:09:26,768 | ERROR | read (Bundle 82) | ThingManager                     | 115 - org.eclipse.smarthome.core.thing - 0.8.0.201510151608 | Exception occured while calling thing handler factory 'org.ec
lipse.smarthome.binding.yahooweather.internal.YahooWeatherHandlerFactory@1a75fa79': null
java.lang.NullPointerException
        at org.eclipse.smarthome.binding.yahooweather.handler.YahooWeatherHandler.startAutomaticRefresh(YahooWeatherHandler.java:101)[177:org.eclipse.smarthome.binding.yahooweather:0.8.0.201510151608]
        at org.eclipse.smarthome.binding.yahooweather.handler.YahooWeatherHandler.initialize(YahooWeatherHandler.java:75)[177:org.eclipse.smarthome.binding.yahooweather:0.8.0.201510151608]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:116)[115:org.eclipse.smarthome.core.thing:0.8.0.201510151608]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:502)[115:org.eclipse.smarthome.core.thing:0.8.0.201510151608]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:1)[115:org.eclipse.smarthome.core.thing:0.8.0.201510151608]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_60]
2015-10-15 23:09:26,770 | INFO  | ESH-safeCall-3   | ThingStatusInfoEvent             | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' updated: UNINITIALIZED (HANDLER
_INITIALIZING_ERROR)
2015-10-15 23:09:26,779 | INFO  | pool-28-thread-1 | FeaturesServiceImpl              | 9 - org.apache.karaf.features.core - 4.0.2 |   org.openhab.binding.avmfritz/2.0.0.201510152107
2015-10-15 23:09:26,779 | INFO  | ESH-safeCall-3   | ThingStatusInfoChangedEvent      | 120 - org.eclipse.smarthome.io.monitor - 0.8.0.201510151608 | 'yahooweather:weather:berlin' changed from ONLINE to UNINITIA
LIZED (HANDLER_INITIALIZING_ERROR)
</pre>

**Comment by maggu2810** [comment]
If you will apply my karaf-master PR (#435), I rebased your PR (#425) to the karaf-master branch.
https://github.com/maggu2810/openhab2/tree/karaf-demo-feature

**Comment by kaikreuzer** [comment]
Thanks, I merged #435 and created a new PR based on your rebased branch: https://github.com/openhab/openhab2/pull/437.

**Comment by kaikreuzer** [comment]
> But I do not see any restarts in the logs.

Even not for the demo feature?

> I see an exception thrown by the yahoo weather binding

I hope that this will be fixed by https://github.com/eclipse/smarthome/pull/491

**Comment by maggu2810** [comment]
Will try it with increased level severity.
But I need some time (very busy ATM).

**Comment by maggu2810** [comment]
Perhaps I found a bottle neck.
We are installing bundles that needs to be resolved by Maven.
If you set the log level to debug, you see that all the repos are connected.

<pre>
2015-10-16 14:19:19,483 | DEBUG | dataResolver-4-1 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection request: [route: {s}->https://oss.sonatype.org:443][total kept alive: 0; route allocated: 0 of 20; total allocated: 0 of 40]
2015-10-16 14:19:19,483 | DEBUG | dataResolver-4-0 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection request: [route: {}->http://repository.apache.org:80][total kept alive: 0; route allocated: 0 of 20; total allocated: 0 of 40]
2015-10-16 14:19:19,484 | DEBUG | dataResolver-4-2 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection request: [route: {s}->https://oss.sonatype.org:443][total kept alive: 0; route allocated: 0 of 20; total allocated: 0 of 40]
2015-10-16 14:19:19,484 | DEBUG | dataResolver-4-1 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection leased: [id: 16][route: {s}->https://oss.sonatype.org:443][total kept alive: 0; route allocated: 2 of 20; total allocated: 3 of 40]
2015-10-16 14:19:19,484 | DEBUG | dataResolver-4-1 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Opening connection {s}->https://oss.sonatype.org:443
2015-10-16 14:19:19,484 | DEBUG | dataResolver-4-1 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connecting to oss.sonatype.org/54.88.69.55:443
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-3 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection request: [route: {}->http://repository.springsource.com:80][total kept alive: 0; route allocated: 0 of 20; total allocated: 3 of 40]
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-2 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection leased: [id: 18][route: {s}->https://oss.sonatype.org:443][total kept alive: 0; route allocated: 2 of 20; total allocated: 3 of 40]
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-0 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection leased: [id: 17][route: {}->http://repository.apache.org:80][total kept alive: 0; route allocated: 1 of 20; total allocated: 3 of 40]
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-2 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Opening connection {s}->https://oss.sonatype.org:443
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-0 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Opening connection {}->http://repository.apache.org:80
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-3 | olingHttpClientConnectionManager | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection leased: [id: 19][route: {}->http://repository.springsource.com:80][total kept alive: 0; route allocated: 1 of 20; total allocated: 4 of 40]
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-3 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Opening connection {}->http://repository.springsource.com:80
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-0 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connecting to repository.apache.org/140.211.11.112:80
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-2 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connecting to oss.sonatype.org/54.88.69.55:443
2015-10-16 14:19:19,488 | DEBUG | dataResolver-4-3 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connecting to repository.springsource.com/54.231.8.209:80
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection established 192.168.0.161:52032<->54.231.8.209:80
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Executing request GET /maven/bundles/external/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Target auth state: UNCHALLENGED
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Proxy auth state: UNCHALLENGED
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> GET /maven/bundles/external/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Cache-control: no-cache
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Cache-store: no-store
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Pragma: no-cache
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Expires: 0
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Accept-Encoding: gzip
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> User-Agent: Aether
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> User-Agent: Aether
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Host: repository.springsource.com
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> Connection: Keep-Alive
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "GET /maven/bundles/external/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Cache-control: no-cache[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Cache-store: no-store[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Pragma: no-cache[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Expires: 0[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Accept-Encoding: gzip[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Host: repository.springsource.com[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "Connection: Keep-Alive[\r][\n]"
2015-10-16 14:19:19,624 | DEBUG | dataResolver-4-3 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-19 >> "[\r][\n]"
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection established 192.168.0.161:47506<->140.211.11.112:80
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Executing request GET /content/groups/snapshots-group/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Target auth state: UNCHALLENGED
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Proxy auth state: UNCHALLENGED
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> GET /content/groups/snapshots-group/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Cache-control: no-cache
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Cache-store: no-store
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Pragma: no-cache
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Expires: 0
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Accept-Encoding: gzip
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> User-Agent: Aether
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> User-Agent: Aether
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Host: repository.apache.org
2015-10-16 14:19:19,691 | DEBUG | dataResolver-4-0 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> Connection: Keep-Alive
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "GET /content/groups/snapshots-group/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Cache-control: no-cache[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Cache-store: no-store[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Pragma: no-cache[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Expires: 0[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Accept-Encoding: gzip[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Host: repository.apache.org[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "Connection: Keep-Alive[\r][\n]"
2015-10-16 14:19:19,692 | DEBUG | dataResolver-4-0 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-17 >> "[\r][\n]"
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | HttpClientConnectionOperator     | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Connection established 192.168.0.161:53668<->54.88.69.55:443
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Executing request GET /content/repositories/snapshots/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Target auth state: UNCHALLENGED
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | MainClientExec                   | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | Proxy auth state: UNCHALLENGED
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> GET /content/repositories/snapshots/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Cache-control: no-cache
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Cache-store: no-store
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Pragma: no-cache
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Expires: 0
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Accept-Encoding: gzip
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> User-Agent: Aether
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> User-Agent: Aether
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Host: oss.sonatype.org
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | headers                          | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> Connection: Keep-Alive
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "GET /content/repositories/snapshots/org/eclipse/smarthome/binding/org.eclipse.smarthome.binding.hue/0.8.0-SNAPSHOT/maven-metadata.xml HTTP/1.1[\r][\n]"
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Cache-control: no-cache[\r][\n]"
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Cache-store: no-store[\r][\n]"
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Pragma: no-cache[\r][\n]"
2015-10-16 14:19:19,767 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Expires: 0[\r][\n]"
2015-10-16 14:19:19,768 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Accept-Encoding: gzip[\r][\n]"
2015-10-16 14:19:19,768 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,768 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "User-Agent: Aether[\r][\n]"
2015-10-16 14:19:19,768 | DEBUG | dataResolver-4-1 | wire                             | 4 - org.ops4j.pax.logging.pax-logging-api - 1.8.4 | http-outgoing-16 >> "Host: oss.sonatype.org[\r][\n]"
</pre>

**Comment by kaikreuzer** [comment]
That's bad indeed! It should only check the locally bundled repos instead.

**Comment by maggu2810** [comment]
This are just a few lines of the log.

Do you remember:
ATM the content of the features / kar files are not copied to the distribution.

Also for SNAPSHOTS it is normal, that Maven wants to find the newest SNAPSHOT and look ups in specific intervals.
You cannot rely that the SNAPSHOT in your local repo is that one that should be used. So Maven would download every SNAPSHOT only once and you would get never a new one.

As already posted to ESH bugzilla, it would be great to get dated / versionized artifacts.

**Comment by maggu2810** [comment]
First we need a decision if we want to add all the content of the kar / features to the distribution, also if it is never installed. This would increase the size of the distribution a lot.
So all the stripping bundles to fit to a small size only suffers the need of RAM, not of persistent storage.

Next step would be to use ESH milestones, to use a fixed reference e.g. for openHAB 2 beta.

**Comment by kaikreuzer** [comment]
> First we need a decision if we want to add all the content of the kar / features to the distribution

Actually, I thought we did that already and I just didn't yet find the folder where they are in ;-)
So yes, that is what I expect. If there are requests from people that require a smaller distribution, this could potentially be a second option to offer then.

> Next step would be to use ESH milestones, to use a fixed reference e.g. for openHAB 2 beta.

Is it not possible to have snapshots in a kar and "detach" the distribution from theremote Maven repo?

**Comment by maggu2810** [comment]
For the openhab2-runtime distribution of the karaf branch the bundled repo is located using "runtime/karaf/system/".

Sure, you could drop all remote repos.
But then you could install no feature that is not part of the distribution.
You could also have a look at "runtime/karaf/etc/org.ops4j.pax.url.mvn.cfg" e.g. "org.ops4j.pax.url.mvn.defaultRepositories" and "org.ops4j.pax.url.mvn.repositories".

For me one of the decision to use Karaf has been, that I can install stuff later without bundle them at build time.

**Comment by kaikreuzer** [comment]
> For me one of the decision to use Karaf has been, that I can install stuff later without bundle them at build time.

No, for me the openHAB distro is about openHAB and not about "out-of-the-box" installing of any unrelated software. As long as we do not maintain any remote repo for openHAB updates or additional addons, which are not packaged in the distro, I do not see any reason to have remote repos configured.

And for users who really want to install other stuff, they are free to configure Karaf how they like - i.e. they can themselves easily add any repo to org.ops4j.pax.url.mvn.repositories.

**Comment by maggu2810** [comment]
No problem at all.

I just tried to suffer both requirements (yours and mine) in one distribution (the upstream one). To reduce the work and combine our efforts.

But it seems that I cannot use this offline runtime for my private / spare time projects.

For your needs an offline distribution will be fine.

Just tried to solve this without an other dist. runtime.

**Comment by kaikreuzer** [comment]
Hope this does not mean too much additional effort for you, but yes, I think we should not mix those requirements here.

**Comment by maggu2810** [comment]
The additional effort are the files that are part of the "standard" Karaf distribution and needs to be changed. On every Karaf bump we have to look what has been changed on the original files and adapt our changes (e.g. do our changes again on the bumped files).
This would be easier if this mechanism is applied to the Karaf maven plugin: https://github.com/apache/karaf/pull/80
Perhaps there is another maven plugin that could be used for this mechansim already.

**Comment by kaikreuzer** [comment]
I can live with having to sync the one file for the moment when upgrading to a newer Karaf version. I have created https://github.com/openhab/openhab2/pull/450 and with that, the installation of the demo feature is indeed much faster.
